subworkflows/local/qc.nf

```diff
@@ -0,0 +1,275 @@
+import { GENERATE_MISS_HET_PLOT } from "../../modules/qc_report/generate_miss_het_plot.nf"
+import { GENERATE_SNP_MISSINGNESS_PLOT } from "../../modules/qc_report/generate_snp_missingness_plot.nf"
+import { IN_MD5 } from "../../modules/qc_report/in_md5.nf"
+import { OUT_MD5 } from "../../modules/qc_report/out_md5.nf"
+import { PRODUCE_REPORTS } from "../../modules/qc_report/produce_reports.nf"
+
+
+//======================
+// Definitions of local workflows
+//======================
+
+//TODO Restructure this logic after the QC_PROCESSES workflow is done
+workflow QC_INPUT_VALIDATION {
+
+if (params.case_control) {
+  ccfile = params.case_control
+  Channel.fromPath(ccfile).into { cc_ch; cc2_ch }
+  col    = params.case_control_col
+  diffpheno = "--pheno cc.phe --pheno-name $col"
+  if (params.case_control.toString().contains("s3://") || params.case_control.toString().contains("az://")) {
+       println "Case control file is in the cloud so we can't check it"
+  } else
+  if (! file(params.case_control).exists()) {
+     error("\n\nThe file <${params.case_control}> given for <params.case_control> does not exist")
+    } else {
+      def line
+      new File(params.case_control).withReader { line = it.readLine() }
+      fields = line.split()
+      if (! fields.contains(params.case_control_col))
+	  error("\n\nThe file <${params.case_control}> given for <params.case_control> does not have a column <${params.case_control_col}>\n")
+    }
+
+} else {
+  diffpheno = ""
+  col = ""
+  cc_ch  = Channel.value.into("none").into { cc_ch; cc2_ch }
+}
+
+
+phenotype_ch = getSubChannel(params.phenotype,"pheno",params.pheno_col)
+batch_ch     = getSubChannel(params.batch,"batch",params.batch_col)
+
+
+
+
+/* Define the command to add for plink depending on whether sexinfo is
+ * available or not.
+ */
+
+if ( nullfile.contains(params.sexinfo_available) ) {
+  sexinfo = "--allow-no-sex"
+  extrasexinfo = ""
+  println "Sexinfo not available, command --allow-no-sex\n"
+} else {
+  sexinfo = ""
+  extrasexinfo = "--must-have-sex"
+  println "Sexinfo available command"
+}
+
+
+/* Get the input files -- could be a glob
+ * We match the bed, bim, fam file -- order determined lexicographically
+ * not by order given, we check that they exist and then
+ * send the all the files to raw_ch and just the bim file to bim_ch */
+inpat = "${params.input_dir}/${params.input_pat}"
+
+
+
+if (inpat.contains("s3://") || inpat.contains("az://")) {
+  print "Here"
+  this_checker = { it -> return it}
+} else {
+  this_checker = checker
+}
+
+
+    Channel
+   .fromFilePairs("${inpat}.{bed,bim,fam}",size:3, flat : true)
+    { file -> file.baseName }  \
+      .ifEmpty { error "No matching plink files" }        \
+      .map { a -> [this_checker(a[1]), this_checker(a[2]), this_checker(a[3])] }\
+      .multiMap {  it ->
+         raw_ch: it
+         bim_ch: it[1]
+         inpmd5ch : it
+       }.set {checked_input}
+
+}
+
+
+workflow QC_PROCESSES {
```

Comment:
@Mxrcon , I've tried to simplify the overall implementation design and have centered it around the 4 channels such as `qc1_ch` ... `qc4_ch`.

I request you to please have a look at the `QC_PROCESSES` workflow and compare it with the DAG here https://github.com/nf-core/gwas/issues/29#issuecomment-1160037132 ?


The code within the `QC_PROCESSES` itself is pseudo-Nextflow at the moment and relies on the `pipes` concept.

I have a feeling that there must be a few pathways which I might have missed but would appreciate a quick review from you to make sure that there isn't anything obvious I am missing out, before taking it forward.

Reply:
@abhi18av This workflow seems to be going in the right direction, at least I felt that every pathway is inserted on it. maybe we can double check with sanity tests after a stub run, what do you think? Also what are the next steps on the workflow deployment? how can I be useful?

Reply:
@Mxrcon , I think that now we've reached a stage where we can effectively test our changes https://github.com/nf-core/gwas/pull/32#issuecomment-1204796864 

I'll share my dev workflow with you when we connect next and then we can continue working w.r.t the templates and finalize the QC workflow.